webapp/tests/karma/unit/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+describe('Telemetry service', () => {
+  'use strict';
+
+  const NOW = new Date(2018, 10, 10, 12, 33).getTime();
+
+  let clock,
+    service,
+    $window,
+    DB,
+    storageGetItem,
+    storageSetItem,
+    storageRemoveItem,
+    pouchDb;
+
+  beforeEach(() => {
+    module('inboxApp');
+    clock = sinon.useFakeTimers(NOW);
+
+    storageGetItem = sinon.stub();
+    storageSetItem = sinon.stub();
+    storageRemoveItem = sinon.stub();
+
+    pouchDb = {};
+
+    $window = {
+      localStorage: {
+        getItem: storageGetItem,
+        setItem: storageSetItem,
+        removeItem: storageRemoveItem,
+      },
+      navigator: {},
+      PouchDB: () => pouchDb,
+    };
+
+    DB = {
+      info: sinon.stub(),
+      put: sinon.stub(),
+    };
+
+    module($provide => {
+      $provide.value('$q', Q);
+      $provide.value('$window', $window);
+      $provide.factory('DB', KarmaUtils.mockDB(DB));
+      $provide.value('Session', {
+        userCtx: () => {
+          return { name: 'greg' };
+        },
+      });
+    });
+    inject(_Telemetry_ => {
+      service = _Telemetry_;
+    });
+  });
+
+  afterEach(function() {
+    clock.restore();
+  });
+
+  describe('Record', () => {
+    it('records a piece of telemetry', () => {
+      storageGetItem.withArgs('medic-greg-telemetry-db').returns('dbname');
+      storageGetItem
+        .withArgs('medic-greg-telemetry-date')
+        .returns(Date.now().toString());
+
+      pouchDb.post = sinon.stub().resolves();
+
+      return service.record('test', 100).then(() => {
+        chai.expect(pouchDb.post.callCount).to.equal(1);
+        chai
+          .expect(pouchDb.post.args[0][0])
+          .to.deep.include({ key: 'test', value: 100 });
+        chai.expect(pouchDb.post.args[0][0].date_recorded).to.be.above(0);
+        chai.expect(storageGetItem.callCount).to.equal(2);
```

Comment:
Do we care how many times `localStorage.getItem()` is called, or is this testing an implementation detail?

Reply:
It's a fine line between strife and crime, as they say.

(I mean, IDK, I kind of dislike unit tests in general because they nearly always end up testing implementation details. You could argue this shows that we're retrieving the right things, at least sort of, though to truly test we'd have to check what each parameter was)